File: solutions/day-04/day04Jokot.py
from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:

        mapGroup = defaultdict(list)

        for s in strs:
            key = self.createMapKey(s)
            mapGroup[key].append(s)
        
        return list(mapGroup.values())
            
    def createMapKey(self, str1):
        key = [0] * 26

        for c in str1:
            key[ord(c) - ord('a')] += 1
        
        return tuple(key)

File: solutions/day-04/day04Jokot.kt
class Solution {
    fun groupAnagrams(strs: Array<String>): List<List<String>> {
        val groups = mutableMapOf<List<Int>, MutableList<String>>()

        for (str in strs) {
            val key = createKey(str)

            if (key in groups) {
                groups[key]?.add(str)
            } else {
                groups[key] = mutableListOf(str)
            }
        }

        return groups.values.toList()
    }

    fun createKey(str: String): List<Int> {
        val keys = MutableList(26) { 0 }

        for (c in str) {
            keys[c.toInt() - 'a'.toInt()] += 1
        }

        return keys
    }
}